pkg/engine/lifecycletest/update_plan_test.go

```diff
@@ -1640,3 +1640,136 @@ func TestPlannedUpdateWithDependentDelete(t *testing.T) {
 	assert.NotNil(t, snap)
 	assert.Nil(t, res)
 }
+
+// TestResourcesTargeted checks that a plan created with targets specified captures only those targets.
+// It checks that trying to constrcut a new resource that was not targeted in the plan fails and that the
+// update with the same --targets specified is compatible with the plan (roundtripped).
+func TestResoucesTargeted(t *testing.T) {
+	t.Parallel()
+
+	host := func() plugin.Host {
+		loaders := []*deploytest.ProviderLoader{
+			deploytest.NewProviderLoader("pkgA", semver.MustParse("1.0.0"), func() (plugin.Provider, error) {
+				return &deploytest.Provider{}, nil
+			}),
+		}
+
+		program := deploytest.NewLanguageRuntime(func(_ plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+			_, _, _, err := monitor.RegisterResource("pkgA:m:typA", "resA", true, deploytest.ResourceOptions{
+				Inputs: resource.PropertyMap{
+					"foo": resource.NewStringProperty("bar"),
+				},
+			})
+			assert.NoError(t, err)
+
+			_, _, _, err = monitor.RegisterResource("pkgA:m:typA", "resB", true, deploytest.ResourceOptions{
+				Inputs: resource.PropertyMap{
+					"foo": resource.NewStringProperty("bar"),
+				},
+			})
+			assert.NoError(t, err)
+
+			return nil
+		})
+
+		host := deploytest.NewPluginHost(nil, nil, program, loaders...)
+		return host
+	}()
+
+	p := &TestPlan{}
+
+	project := p.GetProject()
+
+	// Define a function that creates new resource plans as resource plans get mutated during use.
```

Comment:
You can just clone the plan before passing it into the next update, e.g see https://github.com/pulumi/pulumi/pull/12628/files#diff-114d9ebc788ae0f44bb57d805a87fed37c05998882572331967f95f6ab763afeR1638